smk/include/bwa_index_wrapper.smk

```diff
@@ -0,0 +1,226 @@
+# CLUSTER_WORKLOAD_MANAGER with a request to run only on that node.
+# If it is not run with '--cluster', by mistake, this could cause trouble: now all jobs are run
+# on the same node where you invoked your snakemake command. And this node could wrongly create
+# a 'status' file that denotes sync has been done on a different node. We have a simple solution for this.
+# The node names in CLUSTER_NODES should match the node name accessible by the command 'hostname --short'.
+# The sync status files are named with the local node name. You ask for sync'ing in nodeX, and
+# only nodeX can make that file. This avoids problems when this is NOT run using --cluster.
+# There is anothe problem that multiple sync rules may run the same 'rsync source dest' command
+# on a single server. This would likely cause a race condition. To avoid that, we create a lock
+# using 'flock' so that only one rsync is running for a specific bwa index sync on one node.
+# As a bonus, this also allows rsyncing an NFS bwa index onto a local disk when running without
+# --cluster, if you define CLUSTER_NODES with just your exec node's name.
+################################################################################################
+
+if CLUSTER_WORKLOAD_MANAGER is not None:
+    known_managers = ['SLURM', 'PBS', 'TORQUE', 'SGE', 'LSF']
+    if CLUSTER_WORKLOAD_MANAGER.upper() not in known_managers:
+        raise Exception(f"MIntO error: Unexpected value: CLUSTER_WORKLOAD_MANAGER='{CLUSTER_WORKLOAD_MANAGER}'. Must be one of {known_managers}")
+
+if CLUSTER_NODES is not None:
+
+    # Make sure other variables are defined
+    if CLUSTER_WORKLOAD_MANAGER is None or CLUSTER_LOCAL_DIR is None:
+        raise Exception(f"MIntO error: CLUSTER_WORKLOAD_MANAGER and CLUSTER_LOCAL_DIR must be defined, if CLUSTER_NODES is defined")
+
+    # Baseline    : 5 GB
+    # Size-based  : 22 byte per byte file size
+    # New attempts: +40 GB each time
+    rule BWA_index_contigs_nfs:
+        input:
+            fasta="{somewhere}/{something}.{fasta}"
+        output:
+            "{somewhere}/BWA_index/{something}.{fasta}.0123",
+            "{somewhere}/BWA_index/{something}.{fasta}.amb",
+            "{somewhere}/BWA_index/{something}.{fasta}.ann",
+            "{somewhere}/BWA_index/{something}.{fasta}.bwt.2bit.64",
+            "{somewhere}/BWA_index/{something}.{fasta}.pac",
+        log:
+            "{somewhere}/BWA_index/BWA_index.{something}.{fasta}.log"
+        wildcard_constraints:
+            fasta='fasta|fna'
+        resources:
+            mem = lambda wildcards, input, attempt: 5 + int(22*input.size_mb/1024) + 40*(attempt-1),
+        threads: 4
+        conda:
+            config["minto_dir"]+"/envs/MIntO_base.yml" #bwa-mem2
+        shell:
+            """
+            outdir=$(dirname {output[0]})
+            umask 002 && mkdir -p {CLUSTER_LOCAL_DIR}/$outdir
+            cd {CLUSTER_LOCAL_DIR}/$outdir
```

Comment:
Aren't both `{CLUSTER_LOCAL_DIR}` and `$outdir` full paths?

Reply:
Yes, and they are combined this way deliberately. This is an easy way of making a directory under `CLUSTER_LOCAL_DIR` that wouldn't already exist across all nodes. And the entire path is made with group-writable permissions, so that if we share a `site/cluster_def.py` file across the group, any new creation by other group members does not cause permission problems. 

Since all nodes mount the same NFS, this path is unique for a specific project-analysis pair, and cannot be created by different instances of snakemake or between competing rules of the same snakemake instance. Imagine two instances running a rule creating a `temp` directory under `CLUSTER_LOCAL_DIR` on two servers: there is a tiny chance that they both come up with the same name for the `temp` directory.

Reply:
I suspected that this was your reasoning. I'm still concerned about it being confusing for users.